akka-distributed-data/src/main/scala/akka/cluster/ddata/ReplicatedData.scala

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.cluster.ddata
+
+import akka.cluster.UniqueAddress
+
+/**
+ * Interface for implementing a state based convergent
+ * replicated data type (CvRDT).
+ *
+ * ReplicatedData types must be serializable with an Akka
+ * Serializer. It is highly recommended to implement a serializer with
+ * Protobuf or similar. The built in data types are marked with
+ * [[ReplicatedDataSerialization]] and serialized with
+ * [[akka.cluster.ddata.protobuf.ReplicatedDataSerializer]].
+ *
+ * Serialization of the data types are used in remote messages and also
+ * for creating message digests (SHA-1) to detect changes. Therefore it is
+ * important that the serialization produce the same bytes for the same content.
+ * For example sets and maps should be sorted deterministically in the serialization.
+ *
+ * ReplicatedData types should be immutable, i.e. "modifying" methods should return
+ * a new instance.
+ */
+trait ReplicatedData {
+  type T <: ReplicatedData
+
+  /**
+   * Monotonic merge function.
+   */
+  def merge(that: T): T
+
+}
+
+/**
+ * Java API: Interface for implementing a [[ReplicatedData]] in
+ * Java.
+ */
+abstract class AbstractReplicatedData extends ReplicatedData {
+  // it is not possible to use a more strict type, because it is erased somehow, and
+  // the implementation is anyway required to implement
+  // merge(that: ReplicatedData): ReplicatedData
+  type T = AbstractReplicatedData
+
+}
```

Comment:
interesting, then it would be something like:

    abstract class AbstractReplicatedData[D <: AbstractReplicatedData[D]] extends ReplicatedData

but I could not immediately get that to work (but I think that is because of T in ReplicatedData, perhaps we could use this pattern for both java and scala.

Please create a ticket and we can explore it.